padertorch/contrib/examples/speaker_classification/supervised/data.py

```diff
@@ -0,0 +1,118 @@
+
+    ds = ds.map(prepare_example)
+
+    speaker_encoder = LabelEncoder(
+        label_key='speaker_id', storage_dir=storage_dir, to_array=True
+    )
+    speaker_encoder.initialize_labels(dataset=ds, verbose=True)
+    ds = ds.map(speaker_encoder)
+
+    train_set, validate_set, test_set = train_test_split(ds)
+    training_data = prepare_dataset(train_set, batch_size, training=True)
+    validation_data = prepare_dataset(validate_set, batch_size, training=False)
+    test_data = prepare_dataset(
+        test_set, batch_size, training=False, prefetch=prefetch
+    )
+    return training_data, validation_data, test_data
+
+
+def train_test_split(dataset, dev_split=0.1,  test_split=0.1, seed=0):
+    r = np.random.RandomState(seed)
+
+    try:
+        num_examples = len(dataset)
+    except TypeError:
+        raise RuntimeError('dataset must be indexable!')
+
+    indices = np.arange(num_examples)
+
+    dev_size = int(num_examples * dev_split)
+    test_size = int(num_examples * test_split)
+
+    test_candidates = r.choice(indices, size=test_size, replace=False)
+    indices = np.delete(indices, test_candidates)
+    dev_candidates = r.choice(indices, size=dev_size, replace=False)
+    train_candidates = np.delete(indices, dev_candidates)
+
+    return (
+        dataset[train_candidates], dataset[dev_candidates],
+        dataset[test_candidates]
+    )
+
+
+def _prepare_features(example):
+    audio_data = pb.io.load_audio(
+        example['audio_path'], expected_sample_rate=16000
+    )
+    stft_transform = pb.transform.STFT(
+        shift=160, window_length=400, size=512, fading=None, pad=False
+    )
+    stft = stft_transform(audio_data)
+    mel_transform = pb.transform.module_fbank.MelTransform(
+        sample_rate=16000, fft_length=512, n_mels=64, fmin=50
+    )
+    example['mel_transform'] = mel_transform(np.abs(stft) ** 2)
+    return example
+
+
+def _collate_example(example):
+    example = pt.data.utils.collate_fn(example)
+    example['features'] = pad_sequence(
+        example['features'], batch_first=True
+    ).transpose(-1, -2)
+    example['speaker_id'] = np.stack(example['speaker_id'])
+    return example
+
+
+def prepare_dataset(dataset, batch_size=16, training=False, prefetch=True):
+    dataset = dataset.map(_prepare_features)
+
+    def finalize(example):
+        _example = {
+            'example_id': example['example_id'],
+            'features': torch.from_numpy(
+                example['mel_transform'].astype(np.float32)
+            ),
+            'seq_len': example['mel_transform'].shape[-2],
+            'speaker_id': example['speaker_id'].astype(np.int)
+        }
+        if not training:
+            _example['audio_path'] = example['audio_path']
+        return _example
+
+    dataset = dataset.map(finalize)
+
+    if training:
+        dataset = dataset.shuffle(reshuffle=True)
+    if prefetch:
+        dataset = dataset.prefetch(
+            num_workers=8, buffer_size=10*batch_size
```

Comment:
160 is a huge bugger size. Is the data so unbalanced, that `batch_dynamic_time_series_bucket` needs such a large buffer size?

Reply:
I took the value from @JanekEbb and did not experiment with other values. However, the dataset looks unbalanced to me (plot from padercontrib.database.get_bucket_boundaries) :
<img width="421" alt="Bildschirmfoto 2020-09-21 um 12 43 26" src="https://user-images.githubusercontent.com/17953082/93773939-f3f08280-fc20-11ea-882d-538174fa4a29.png">

Reply:
Ah, ok. I just wanted to ensure, it is intended. Depending on the size of the examples, you could run out of memory.